### PR TITLE
Node-based @deprecated checks

### LINKED ISSUE #39212: document.createElement always marked deprecated
**TypeScript Version:**  4.0.0-dev-20200622

**Search Terms:**
- deprecated

**Code**
For the JS
```js
document.createElement('a');
```

**Bug:**

`createElement` is always rendered as deprecated

Screen Shot 2020-06-23 at 1 42 05 PM

I believe this is because one of the `createElement` overloads is deprecated

/cc @Kingwl @DanielRosenwasser We will need to fix this before we can ship `@deprecated` support in the stable version of VS Code

### PR DESCRIPTION
**Pull request by sandersn**
Switch the checker to syntactic checks for `@deprecated` on declarations. This requires a bit more checking of declarations in the
checker at times, but it:

1. Gets rid of work, and a symbol flag, in the binder.
2. Skips work in the checker unless there is a `@deprecated` tag.
3. Makes it fairly simple to only issue errors on particular signatures of overloaded functions.

Fixes #39212, alternative to #39221 -- and tests are based on #39221's.

Edit: This now fixes part of #39374; JSX elements, decorators and tagged template functions are all calls and were not being marked deprecated.

### CODE REVIEW

**Comment by sandersn** [comment]
@typescript-bot perf test this

**Comment by sandersn** [comment]
Note: `isTypeDeclaration` from #39122 is better and should be copied here. I think the tests from here have a few more cases than those in #39122 and should be copied there.

**Comment by Kingwl** [comment]
@typescript-bot pack this.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13274,13 +13274,19 @@ namespace ts {
                         undefined;
         }
 
+        function isUncalledFunctionReference(node: Node, symbol: Symbol) {
+            return !(symbol.flags & SymbolFlags.Function)
```
Comment (Kingwl) [inline src/compiler/checker.ts:13278]: Maybe Method or Constructor?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -13274,13 +13274,19 @@ namespace ts {
                         undefined;
         }
 
+        function isUncalledFunctionReference(node: Node, symbol: Symbol) {
+            return !(symbol.flags & SymbolFlags.Function)
+                || !isCallExpression(findAncestor(node, n => !isAccessExpression(n)) ?? node.parent)
```
Comment (Kingwl) [inline src/compiler/checker.ts:13279]: NewExpression?

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:13279]: Or an invocationexpression in general

Just switch to `||`, it's not a primitive type

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -25013,7 +25018,7 @@ namespace ts {
                 propType = indexInfo.type;
             }
             else {
-                if (prop.flags & SymbolFlags.Deprecated) {
+                if (prop.valueDeclaration?.flags & NodeFlags.Deprecated && isUncalledFunctionReference(node, prop)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:25021]: This is potentially `undefined & NodeFlags.Deprecated` - is that intentional?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -13274,13 +13274,19 @@ namespace ts {
...
+        }
+
         function getPropertyTypeForIndexType(originalObjectType: Type, objectType: Type, indexType: Type, fullIndexType: Type, suppressNoImplicitAnyError: boolean, accessNode: ElementAccessExpression | IndexedAccessTypeNode | PropertyName | BindingName | SyntheticExpression | undefined, accessFlags: AccessFlags) {
             const accessExpression = accessNode && accessNode.kind === SyntaxKind.ElementAccessExpression ? accessNode : undefined;
             const propName = accessNode && isPrivateIdentifier(accessNode) ? undefined : getPropertyNameFromIndex(indexType, accessNode);
             if (propName !== undefined) {
                 const prop = getPropertyOfType(objectType, propName);
                 if (prop) {
-                    if (accessNode && prop.flags & SymbolFlags.Deprecated) {
+                    if (accessNode && prop.valueDeclaration?.flags & NodeFlags.Deprecated && isUncalledFunctionReference(accessNode, prop)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:13289]: This is potentially `undefined & NodeFlags.Deprecated` - is that intentional?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -22043,9 +22049,8 @@ namespace ts {
             const localOrExportSymbol = getExportSymbolOfValueSymbolIfExported(symbol);
             let declaration: Declaration | undefined = localOrExportSymbol.valueDeclaration;
 
-            const target = (symbol.flags & SymbolFlags.Alias ? resolveAlias(symbol) : symbol);
-            if (target.flags & SymbolFlags.Deprecated) {
-                errorOrSuggestion(/* isError */ false, node, Diagnostics._0_is_deprecated, node.escapedText as string);
+            if (declaration?.flags & NodeFlags.Deprecated && isUncalledFunctionReference(node.parent, localOrExportSymbol)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:22052]: This is potentially `undefined & NodeFlags.Deprecated` - is that intentional?

**Comment by sandersn** [comment]
@typescript-bot perf test this

**Comment by sandersn** [comment]
@typescript-bot perf test this

**Comment by sandersn** [comment]
Overall, there's still a good bit of variation in our perf tests. But the perf on this is still pretty good even after supporting JSX et al. I'm going to merge this since I've addressed all the comments on it and in unblocks #39388.

**Pull request merged**